pkg/vault/test/authorizer/AuthorizerWithAdaptorValidation.test.ts

```diff
@@ -0,0 +1,123 @@
+import { expect } from 'chai';
+import { ethers } from 'hardhat';
+import { Contract } from 'ethers';
+import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers';
+import { sharedBeforeEach } from '@balancer-labs/v2-common/sharedBeforeEach';
+import { deploy } from '@balancer-labs/v2-helpers/src/contract';
+import { ANY_ADDRESS, ZERO_ADDRESS } from '@balancer-labs/v2-helpers/src/constants';
+import { impersonateAccount, setBalance } from '@nomicfoundation/hardhat-network-helpers';
+import { fp } from '@balancer-labs/v2-helpers/src/numbers';
+import { actionId } from '@balancer-labs/v2-helpers/src/models/misc/actions';
+
+describe('AuthorizerWithAdaptorValidation', () => {
+  let vault: Contract, authorizerAdaptor: Contract, adaptorEntrypoint: Contract;
+  let user: SignerWithAddress, other: SignerWithAddress;
+  let authorizer: Contract;
+  let actualAuthorizer: Contract;
+  let admin: SignerWithAddress;
+
+  before('setup signers', async () => {
+    [, admin, user, other] = await ethers.getSigners();
+  });
+
+  sharedBeforeEach('deploy old authorizer and helper', async () => {
+    actualAuthorizer = await deploy('MockBasicAuthorizer', { from: admin });
+
+    vault = await deploy('Vault', { args: [actualAuthorizer.address, ZERO_ADDRESS, 0, 0] });
+
+    authorizerAdaptor = await deploy('v2-liquidity-mining/AuthorizerAdaptor', {
+      args: [vault.address],
+    });
+
+    adaptorEntrypoint = await deploy('v2-liquidity-mining/AuthorizerAdaptorEntrypoint', {
+      args: [authorizerAdaptor.address],
+    });
+
+    authorizer = await deploy('AuthorizerWithAdaptorValidation', {
+      args: [actualAuthorizer.address, authorizerAdaptor.address, adaptorEntrypoint.address],
+    });
+  });
+
+  it('stores the actual (existing basic) authorizer', async () => {
+    expect(await authorizer.getActualAuthorizer()).to.equal(actualAuthorizer.address);
+  });
+
+  it('stores the authorizer adaptor', async () => {
+    expect(await authorizer.getAuthorizerAdaptor()).to.equal(authorizerAdaptor.address);
+  });
+
+  it('stores the authorizer adaptor entrypoint', async () => {
+    expect(await authorizer.getAuthorizerAdaptorEntrypoint()).to.equal(adaptorEntrypoint.address);
+  });
+
+  describe('canPerform', () => {
+    const ROLE_1 = '0x0000000000000000000000000000000000000000000000000000000000000001';
+    const ROLE_2 = '0x0000000000000000000000000000000000000000000000000000000000000002';
+
+    let adaptorSigner: SignerWithAddress;
+
+    sharedBeforeEach('grant permission on the old authorizer', async () => {
+      await actualAuthorizer.connect(admin).grantRole(ROLE_1, user.address);
+    });
+
+    sharedBeforeEach('impersonate adaptor', async () => {
+      await impersonateAccount(authorizerAdaptor.address);
+      await setBalance(authorizerAdaptor.address, fp(1));
+
+      // Simulate a call from the real AuthorizerAdaptor by "casting" it as a Signer,
+      // so it can be used with `connect` like an EOA
```

Comment:
```suggestion
```

This is already covered in the block description, I don't think we need to add much